contrib/hualloc/hu_alloc.h

```diff
@@ -0,0 +1,1615 @@
+#include <Windows.h>
+
+#define PERTHREAD __declspec(thread)
+#define NOINLINE __declspec(noinline)
+#define THREAD_FUNC_RETURN DWORD
+#define SchedYield SwitchToThread
+
+#else
+#include <immintrin.h>
+#include <pthread.h>
+#include <sys/mman.h>
+#include <sched.h>
+#include <unistd.h>
+
+#define PERTHREAD __thread
+#define NOINLINE __attribute__((noinline))
+#define THREAD_FUNC_RETURN void*
+#define SchedYield sched_yield
+
+static __inline__ unsigned char __attribute__((__always_inline__, __nodebug__))
+_BitScanForward64(unsigned long *_Index, unsigned long long _Mask)
+{
+    if (!_Mask)
+        return 0;
+    *_Index = __builtin_ctzll(_Mask);
+    return 1;
+}
+static __inline__ unsigned char __attribute__((__always_inline__, __nodebug__))
+_BitScanReverse64(unsigned long *_Index, unsigned long long _Mask)
+{
+    if (!_Mask)
+        return 0;
+    *_Index = 63 - __builtin_clzll(_Mask);
+    return 1;
+}
+
+inline void Sleep(yint x)
+{
+    usleep(x * 1000);
+}
+#endif
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
+#endif
+
+#undef PAGE_SIZE
+// platform specific actually
+#define PAGE_SIZE 4096
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef _win_
+// no huge pages for windows so far
+static void* OsReserve(yint sz)
+{
+    return VirtualAlloc(0, sz, MEM_RESERVE, PAGE_NOACCESS);
+}
+static void OsCommit(void *p, yint sz)
+{
+    VirtualAlloc(p, sz, MEM_COMMIT, PAGE_READWRITE);
+}
+static void OsReclaim(void *p, yint sz)
+{
+    VirtualFree(p, sz, MEM_DECOMMIT);
+}
+static void* OsAlloc(yint *pSz)
+{
+    yint &sz = *pSz;
+    return VirtualAlloc(0, sz, MEM_COMMIT, PAGE_READWRITE);
+}
+static void OsFree(void *p, yint sz)
+{
+    (void)sz;
+    VirtualFree(p, 0, MEM_RELEASE); // size not needed
+}
+
+#else
+static void* OsReserve(yint sz)
+{
+    char *res = 0;
+    for (ui64 rnd = 0x554721eb2148deadull;; rnd = rnd * 0x83e385b21a346273ull + 0x3491274383712345ull) {
+        char *ptr = (char*)(rnd & 0x0fffff000000); // high address bits are zero (48 bits for x64), low bits are zero for alignment
+        res = (char*)mmap(ptr, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON | MAP_NORESERVE, -1, 0);
+        if (res == ptr) {
+            break;
+        }
+    }
+    Y_ASSERT(res);
+#ifdef MADV_HUGEPAGE
+    madvise(res, sz, MADV_HUGEPAGE);
+#endif
+    return res;
+}
+static void OsCommit(void *p, yint sz)
+{
+    // no special action required
+    (void)p;
+    (void)sz;
+}
+static void OsReclaim(void *p, yint sz)
+{
+    //madvise(segment, SEGMENT_SIZE, MADV_DONTNEED); // super slow for no apparent reason, use MAP_FIXED to map new memory over non needed
+    void *res = mmap(p, sz, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_PRIVATE | MAP_ANON | MAP_NORESERVE, -1, 0);
+    if (res != p) {
+        abort(); // MAP_FIXED failed
+    }
+#ifdef MADV_HUGEPAGE
+    madvise(res, sz, MADV_HUGEPAGE);
+#endif
+}
+static void* OsAlloc(yint *pSz)
```

Comment:
I propose to modify the algorithm. Instead of hinting OS that we want huge pages by rounding sz, try to alloc huge pages explicitly and go default route if it fails:

```C
#ifdef MADV_HUGEPAGE
    if (sz > 5 * 1000 * 1000) {
        hugeSz = roundUp(sz, hugePageSize);
        ptr = mmap(0, hugeSz, ... MAP_HUGETLB ...);
        if (ptr) {sz = hugeSz; return ptr;}
    }
#endif
    sz = roundUp(sz, PAGE_SIZE);
....
```

Note: it will have the side effect of more efficient memory usage when huge pages aren't available. And as consequence, it can increase the speed of real memory fragmentation (i.e. decreasing over time the number of huge pages available).

PS: this algorithm works for Windows too and in fact borrowed from my Windows programs. I don't have the slightest idea of intrinsics of any Unix - may be you have reasons to use madvise rather than this approach.